sdk/search/Azure.Search.Documents/src/SearchDocument/SearchDocument.Core.cs

```diff
@@ -0,0 +1,166 @@
+        private protected bool TryGetValue(string name, Type type, out object value)
+        {
+            Argument.AssertNotNullOrEmpty(name, nameof(name));
+            type ??= typeof(object);
+
+            bool found = _values.TryGetValue(name, out value);
+            if (found && type != typeof(object))
+            {
+                value = ConvertValue(value, type);
+            }
+            return found;
+        }
+
+        /// <summary>
+        /// Attempt to convert a value to the desired type.  This is full of
+        /// Search specific conversions and I'd expect to be able to override
+        /// a base version of this provided by DynamicData.
+        /// </summary>
+        /// <param name="value">The value to convert.</param>
+        /// <param name="type">The type to convert to.</param>
+        /// <returns>The converted value.</returns>
+        private protected static object ConvertValue(object value, Type type)
+        {
+            // Short circuit values that are already in the right format
+            if (value == null || type == null || type.IsAssignableFrom(value.GetType()))
+            {
+                return value;
+            }
+
+            // Just unwrap Nullable<>s since we've already handled null above
+            if (type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>))
+            {
+                Type elementType = type.GetGenericArguments()[0];
+                return ConvertValue(value, elementType);
+            }
+
+            // Arrays deserialize as object[] but we usually want T[]
+            if (type.IsArray && type.GetArrayRank() == 1 && value is Array values)
+            {
+                Type elementType = type.GetElementType();
+                int length = values.Length;
+                Array converted = Array.CreateInstance(elementType, length);
+                for (int i = 0; i < length; i++)
+                {
+                    converted.SetValue(
+                        ConvertValue(values.GetValue(i), elementType),
+                        i);
+                }
+                return converted;
+            }
+
+            // Special case DateTime/DateTimeOffset/and textual doubles that we
+            // don't convert until a user requests them in a given format
+            if (value is string text)
+            {
+                if (type == typeof(DateTime))
+                {
+                    return DateTime.Parse(text, CultureInfo.InvariantCulture);
+                }
+                else if (type == typeof(DateTimeOffset))
+                {
+                    return DateTimeOffset.Parse(text, CultureInfo.InvariantCulture);
+                }
+                else if (type == typeof(double))
+                {
+                    if (string.Equals(text, Constants.NanValue, StringComparison.Ordinal))
+                    {
+                        return double.NaN;
```

Comment:
Just to be clear because the quotes matter so much here, `NaN` isn't valid JSON but `"NaN"` is. :smile:  We send `"NaN"` over the wire, leave it in that form inside `SearchDocument`, and this conversion code only runs when you call `doc.GetDouble("foo")` or say `dynamic doc = ...; double foo = doc.foo;`.  If this index's `foo` field was of type `string`, we don't get in trouble by eagerly converting.